[Diagnostics] Don't fix partial mismatch for sub-types associated with…

xedin opened the pull request:
… optional conversion

If mismatch detected by `repairFailures` is related to a complex
wrapped value of optional type formed from optional-to-optional
or value-to-optional conversion let's not try to fix it directly
but let `simplifyRestrictedConstraintImpl` record a top-level fix
for more context.

Resolves: rdar://problem/59703585

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin commented:
@swift-ci please smoke test

LucianoPAlmeida commented on test/Constraints/function_conversion.swift:
```diff
@@ -68,3 +68,15 @@ noEscapeParam = takesAny // expected-error {{converting non-escaping value to 'A
+
+  var cb: Fn? = nil
+
+  cb = swiftCallback
```
I wonder if for `AssignExpr` the diagnostic could be `cannot assign value of type 'A' to 'B'` instead of a `cannot convert value...`? :)

xedin replied:
Ah true, I'll fix later tonight!

xedin replied:
Fixed!

xedin committed:
[Diagnostic] Don't fix partial mismatch for sub-types associated with optional conversion

If mismatch detected by `repairFailures` is related to a complex
wrapped value of optional type formed from optional-to-optional
or value-to-optional conversion let's not try to fix it directly
but let `simplifyRestrictedConstraintImpl` record a top-level fix
for more context.

Resolves: rdar://problem/59703585

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test
